### PR TITLE
[batch] update ubuntu image version to latest

### PR DESCRIPTION
**Pull request by sjparsa**
Upgrade to Ubuntu 22.04 everywhere including for the Batch Worker VMs in Azure and Google.

### CODE REVIEW

**Comment by danking** [comment]
@daniel-goldstein should this PR update line 38 of main.tf which uses Ubuntu 20.04 for the batch worker? Do we need to take special action if that happens?

File: batch/gcp-create-worker-image.sh
Comment (danking) [inline batch/gcp-create-worker-image.sh:24]: @sjparsa , I don't see any dev docs about how to update a worker image. Did someone explain how to do that to you?

I confirmed that batch-worker-13 already exists:
```
(base) dking@wm28c-761 hail % gcloud compute images list --project hail-vdc
NAME                                                  PROJECT              FAMILY                             DEPRECATED  STATUS
...
batch-worker-13                                       hail-vdc                                                            READY
```

File: docker/third-party/images.txt
Comment (danking) [inline docker/third-party/images.txt:19]: Remove focal here as well. I think we can nix all the old ubuntu's. @daniel-goldstein @jigold objections to that?

**Review by danking — CHANGES_REQUESTED** [review]
Looks great, just a couple questions about Azure, a possible more complete cleanup, and a question about docs.

File: docker/third-party/images.txt
Comment (jigold) [reply docker/third-party/images.txt:19]: I think we should leave `20.04`. The others are fine to nix.

File: batch/gcp-create-worker-image.sh
Comment (sjparsa) [reply batch/gcp-create-worker-image.sh:24]: Yes Daniel talked to me about it! I asked him to make batch-worker-13 with the update so I think that should be why it's showing up? @daniel-goldstein

**Comment by daniel-goldstein** [comment]
> @daniel-goldstein should this PR update line 38 of main.tf which uses Ubuntu 20.04 for the batch worker? Do we need to take special action if that happens?
> 
> If we do this we also need to change `infra/azure/create_bootstrap_vm.sh` which references `focal`.

@jigold Is this terraform resource needed in the gallery in order for `az-create-worker-image.sh` to run? It's unclear to me whether `az-create-worker-image.sh` references this shared image.

File: batch/gcp-create-worker-image.sh
Comment (danking) [reply batch/gcp-create-worker-image.sh:24]: @daniel-goldstein can you also create PR adding a dev-doc for how to do this? TY!

**Comment by daniel-goldstein** [comment]
> > @daniel-goldstein should this PR update line 38 of main.tf which uses Ubuntu 20.04 for the batch worker? Do we need to take special action if that happens?
> > If we do this we also need to change `infra/azure/create_bootstrap_vm.sh` which references `focal`.
> 
> @jigold Is this terraform resource needed in the gallery in order for `az-create-worker-image.sh` to run? It's unclear to me whether `az-create-worker-image.sh` references this shared image.

@jigold bump

**Comment by jigold** [comment]
Just to make sure I understand correctly, are you asking whether this resource `--image Canonical:0001-com-ubuntu-server-focal:20_04-lts:latest` needs to be in Terraform? This is an image from the Azure Marketplace. I think you can just pick a different one and it will be fine.

**Comment by daniel-goldstein** [comment]
No sorry I'm asking if [this](https://github.com/hail-is/hail/blob/3a8be2b37ec387cbf0664354148a66e5930f2f88/infra/azure/modules/batch/main.tf#L27-L40) is unused and can be deleted or whether we are using it somewhere.

**Comment by jigold** [comment]
I don't think you can delete it. If you look in the portal here, you'll see it contains all of the boot disk images:
https://portal.azure.com/#@haildev.onmicrosoft.com/resource/subscriptions/22cd45fe-f996-4c51-af67-ef329d977519/resourceGroups/haildev/providers/Microsoft.Compute/galleries/haildev_batch/images/batch-worker/overview

You might need to make a new one image definition for 22.04 -- otherwise, I think you're going to delete our existing images.

**Comment by daniel-goldstein** [comment]
> I don't think you can delete it. If you look in the portal here, you'll see it contains all of the boot disk images: https://portal.azure.com/#@haildev.onmicrosoft.com/resource/subscriptions/22cd45fe-f996-4c51-af67-ef329d977519/resourceGroups/haildev/providers/Microsoft.Compute/galleries/haildev_batch/images/batch-worker/overview
> 
> You might need to make a new, separate image definition for 22.04 -- otherwise, I think you're going to delete our existing images.

Ah, great thanks, this is helpful. I wasn't sure exactly how the script played with that resource. I'll add a new resource for 22.04 and switch over to using that. Once we're fully moved over, we'll follow up with a separate PR that removes the old image.

File: batch/gcp-create-worker-image.sh
Comment (sjparsa) [reply batch/gcp-create-worker-image.sh:27]: Yes just did!

**Review by danking — CHANGES_REQUESTED** [review]
Dismiss when you want me to review.

File: infra/azure/modules/batch/main.tf
Comment (daniel-goldstein) [inline infra/azure/modules/batch/main.tf:57]: I think we had a misunderstanding of what the shared image definition is doing. An image definition does not tie itself to a particular base VM image like Ubuntu 20.04 or Ubuntu 22.04, it just defines a [logical grouping of concrete image versions](https://learn.microsoft.com/en-us/azure/virtual-machines/shared-image-galleries?tabs=azure-cli#image-definitions) with some broad characteristics defined like whether the image versions are Linux or Windows and which hypervisor generation is needed to run them.

This `identifier` block allows us to define a canonical name for the image definition that we are creating, and AFAICT is not intended to reference an existing marketplace identifier. So the existing definition doesn't actually give us a Ubuntu 20.04 base VM image.

Where the base VM image is actually defined is where we create the VM in `az-create-worker-image.sh` with `az vm create` (this makes sense in hindsight because we don't actually reference the image definition in the `az vm create` call so how could it know to pick 20.04 vs 22.04?). In that call we have been using `--image UbuntuTLS`. I think this is Ubuntu 18.04, because that's what I get when running `lsb_release -a` on a production Azure VM. So in this PR I've changed that to be `Ubuntu2204` which seems to have done the trick. I think we still need a different image definition here though because Ubuntu 18.04 and 22.04 need different `hyper_v_generation`. Trying to save an Ubuntu 22.04 VM under the existing definition resulted in an error saying that the VM was incompatible with a V1 image definition.

File: batch/Dockerfile.worker
Comment (danking) [inline batch/Dockerfile.worker:31]: huzzah, at last!

File: docker/third-party/images.txt
Comment (danking) [reply docker/third-party/images.txt:19]: bump: do we need to keep 20.04? Presumably we can kill 18.04 and 19.04.

**Review by danking — DISMISSED** [review]
This is an excellent change. I'm glad we cleaned up the situation in Azure a bit.

File: docker/third-party/images.txt
Comment (daniel-goldstein) [reply docker/third-party/images.txt:19]: See above message, we should keep `ubuntu:20.04` until we migrate the `global-config` then we can delete it.

File: batch/batch/cloud/azure/driver/create_instance.py
Comment (daniel-goldstein) [reply batch/batch/cloud/azure/driver/create_instance.py:291]: Related to the above error.

File: batch/batch/cloud/azure/driver/create_instance.py
Comment (sjparsa) [reply batch/batch/cloud/azure/driver/create_instance.py:280]: Ah unfortunately no I've had to do several database wipes since then so I don't have any of my old stuff..... should I recreate a run without it?

File: batch/batch/cloud/azure/driver/create_instance.py
Comment (sjparsa) [reply batch/batch/cloud/azure/driver/create_instance.py]: Removing this results in errors such as `writing file /sys/fs/cgroup/batch-7-job-1-main/cgroup.procs: No such file or directory`

File: batch/batch/cloud/azure/driver/create_instance.py
Comment (sjparsa) [reply batch/batch/cloud/azure/driver/create_instance.py]: Looks like the :shared wasn't necessary and so I removed it

File: batch/batch/cloud/azure/driver/create_instance.py
Comment (danking) [reply batch/batch/cloud/azure/driver/create_instance.py:274]: @daniel-goldstein  Alright, let's PR that change separately.

File: batch/batch/cloud/azure/driver/create_instance.py
Comment (danking) [reply batch/batch/cloud/azure/driver/create_instance.py:291]: Ah, right, right. This is the worker's VM exposing cgroups to the batch-worker container. Perfect.

File: docker/third-party/images.txt
Comment (danking) [reply docker/third-party/images.txt:19]: And we'll delete 19.04 then too, right?

**Review by danking — APPROVED** [review]
LGTM, two pending threads with Daniel that need not block this PR.

**Pull request merged**
